### PR TITLE
Migration test skeleton

### PR DESCRIPTION
**Pull request by emetsger**
## Summary
Provides the initial testing framework for testing of migrations, and includes an initial test as proof the framework works.  

The PR encompasses the installation and update of migration-related plugins, notably the addition of the `migrate_source_ui` plugin, and an update to the `migrate_source_csv`, which moves to using the `league/csv` library for parsing CSV files.  Novel migration definitions are included in `config/sync` and are baked into the snapshot present in this PR.

Tests are included in the `tests/10-migration-backend-tests` directory, and executed by `tests/10-migration-backend-tests.sh`  Testcafe is used to perform migrations, and Go code is used to verify that the migration succeeded.

Subsequent PRs will add additional migration tests to this framework.

## To Test
1. Invoke `make reset`
1. Login and view the Person Taxonomy, note it is empty.  Note the presence of the 'Migrate Source UI' menu item under the 'Structure' menu.
1. Click on 'Migrate Source UI', and view the list of migrations present in active configuration, including one labeled "Taxonomy: Persons (supports csv)"
1. Invoke `tests/10-migration-backend-tests.sh` and observe that the tests pass.
1. Refresh the Person Taxonomy, note it is populated with three people. 
1. Click on Ansel Adams, note all fields are filled, including birth and death dates, authority source, and that there is a relationship to Edward Weston.
1. Click on Edward Weston and make similar observations.  Note the formatting of Weston's description is clumsy but that is due to the formatting of that information in the CSV file.
1. Feel free to click on Galen Rowell.  There are no relationships in this entry so it is rather unexciting. 

## Bonus
If you're feeling chippy, you can perform the Person migration manually.
1. Invoke `make reset`
1. Navigate to 'Migrate Source UI' 
1. Select "Taxonomy: Persons (supports csv)" from the dropdown
1. Click "Choose File" and navigate your local filesystem to the `tests/10-migration-backend-tests/testcafe/migrations` directory and select `persons-01.csv` and click `Migrate`.
1. Optionally navigate to the Person taxonomy.  Note three People are present, but no relationships exist between Adams and Weston.
1. Click "Choose File" and navigate your local filesystem to the `tests/10-migration-backend-tests/testcafe/migrations` directory and select `persons-02.csv` and click `Migrate`.
1. List the contents of the Person taxonomy.  Note three People are present, complete with the relationships between Adams and Weston. 

A copy of the included `README.md` for the migration test framework:

# IDC Migration Tests

## Invoking the tests

Migration tests may be invoked in isolation by running the `10-migration-backend-tests.sh` script.
> The script should _not_ be invoked from any sub-directories of the `10-migration-backend-tests` directory

The migration tests do participate in the general IDC test framework, and should run automatically when `make test` is invoked.

## How the tests work - an overview

The `10-migration-backend-tests.sh` script is the "controller" of the tests.  It is responsible for executing the various test frameworks and controls the shell exit code.  Each test framework executes in a Docker container, so there are no dependencies or configuration required to perform the tests, except for a working Docker.

The controller script invokes testcafe first, which will perform migrations that result in resources being created in Drupal.  Next, the controller will invoke Go tests which verify the resources were created correctly (e.g. that the data in the migration CSV files are present in the Drupal resources).

Because one test framework creates resources and another test framework verifies the resources, there is unfortunate coupling that may not be readily apparent.  If the CSV files in testcafe are modified, the verification code must almost certainly be updated to account for the changes in the Drupal resources.

### Testcafe performs the migration

The testcafe framework is used to perform a series of migrations.  The _data_ to be migrated resides under the `testcafe` directory in CSV files. The migration _definitions_ are defined in the Drupal active configuration; that is, they're baked in to Drupal already.

The migrations occur just as if someone from the JHU digital collections team were performing a migration.  They must be performed in order so that child resources can reliably refer to their parents.

The Testcafe tests few things
* the presence of the baked in migrations
* the presence of the Migrate Source UI module
* the authorization of an admin user to perform a migration

### Go tests verify the migration

The Go tests are used to verify that the migration performed by testcafe resulted in the creation of the expected Drupal resource.  The Go tests are used to verify the migration definitions baked into Drupal's configuration.

Go verification directly tests things like:
* Did a given piece of data in the migration CSV file get properly translated to a Drupal field
* Were the relationships present in the migration CSV data properly translated to links between Drupal resources

The Go verification tests rely on the Drupal JSONAPI module to retrieve migrated resources, and compare the JSONAPI response with a bespoke JSON format representing the expected response.

### Coupling of test data

As mentioned above, if the migration CSV is updated to add a new field, merge an existing field, or remove a field, or if values change, it is likely that the verification code will need to be updated to accommodate this change.

The CSV test data is present in the `testcafe/migrations` directory, and the Go verification JSON is present in the `verification/expected` directory.  Depending on the nature of the changes to the CSV data, only minor modifications to the verification JSON may be necessary.

However, if the CSV is modified in any non-trivial way (like adding a row, changing a field name or adding a field), the Go verification code itself will need to be updated, along with updated the verification JSON.

### Use of URIs in test data

Drupal does not make it easy to determine the URI of a migrated resource.  Even if the URI of a migrated resource could be determined by the testcafe code, it would be difficult - or at least unorthodox and ungainly - to share the URI of that resource with the verification code.

> To do that properly, a service of some kind would need to be developed which consults the Drupal database, and that's too time-consuming for the moment

Therefore, verification code can't retrieve resources using their URI; instead, it queries for a resource using unique values.  For example, it will retrieve a Person taxonomy term by filtering on the first and last name, or retrieve a node by filtering on its title.

An additional consequence of this is that the migration CSV files _cannot_ refer to resources by their URI.  In a _production_ environment, a migration definition could leverage URIs.  For example, to establish that one Person `schema:knows` another Person, the target of the `schema:knows` relationship could be a URI.  This is made possible by the migration definition expecting and operating on URIs, and by the author of the CSV; the author of the CSV logs into IDC, finds the URI of the target Person, and copies it into the CSV.

However, because URIs are not known in advance in the _test_ environment, relationships like the one illustrated here cannot be established using URIs.  They must be established using some other property, and in this case the migration definitions use _local identifiers_ established by the author of the CSV to create these relationships.

The consequence is that the production environment cannot use URIs to create relationships _and_ have a reliable test for the migration.

### Use of JSONAPI filters

JSONAPI filters are used by the Go verification code to find a single resource based on unique characteristics of the resource.  For example, to filter for a single Person using the JSONAPI, a filter like `filter[name]=Ansel Adams` would be used.  To filter for a Node with a given UUID, a filter like `filter[id]=adb236e4-23d0-4f91-b986-860cad20ed9d`.

The "gotcha" here is that the author of test CSV must bear this in mind when creating test data for migration.  As long as the author uses reasonable values (for example, the string "Moo" is not used as a name for two different Persons, and nodes receive unique titles), this should not be a problem.

As far as I know, there is no easy way to use the JSONAPI to filter for a resource given its URI.

### CODE REVIEW

**Comment by emetsger** [comment]
In its draft form, this PR adds back the PDF.js viewer and Masonry JS libraries.  Those could be removed from this PR if desired.

**Comment by emetsger** [comment]
In order to move this PR out of draft form, the existing Taxonomy migration and test should be updated to align with the current data model.  Tests should pass at that point.

A subsequent PR will flesh out the remaining migrations and tests.

File: tests/10-migration-backend-tests.sh [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+fi
+
+TESTCAFE_TESTS_FOLDER=$(pwd)/testcafe
+
+# Execute migrations using testcafe
+docker run --network gateway -v "${TESTCAFE_TESTS_FOLDER}":/tests testcafe/testcafe --screenshots path=/tests/screenshots,takeOnFails=true chromium /tests/**/*.js
+
+# Verify migrations using go
+
+# Build docker image (TODO: should it be defined in docker-compose.yml to avoid any env issues?)
```
Comment (birkland) [inline tests/10-migration-backend-tests.sh:30]: This is an overarching issue/TODO that needs a good solution not only for migration tests, but all others.

File: tests/10-migration-backend-tests/testcafe/migrations/persons-01.csv [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
...innovative and influential American photographers...\""<a href=\""https://en.wikipedia.org/wiki/Edward_Weston#cite_note-1\"">[1]</a>&nbsp;and \""one of the masters of 20th century photography.\""<a href=\""https://en.wikipedia.org/wiki/Edward_Weston#cite_note-2\"">[2]</a>&nbsp;Over the course of his 40-year career Weston photographed an increasingly expansive set of subjects, including landscapes, still lives, nudes, portraits, genre scenes and even whimsical parodies. It is said that he developed a \""quintessentially American, and especially Californian, approach to modern photography\""<a href=\""https://en.wikipedia.org/wiki/Edward_Weston#cite_note-3\"">[3]</a>&nbsp;because of his focus on the people and places of the American West. In 1937 Weston was the first photographer to receive a&nbsp;<a href=\""https://en.wikipedia.org/wiki/Guggenheim_Fellowship\"">Guggenheim Fellowship</a>, and over the next two years he produced nearly 1,400 negatives using his 8 × 10 view camera. Some of his most famous photographs were taken of the trees and rocks at&nbsp;<a href=\""https://en.wikipedia.org/wiki/Point_Lobos\"">Point Lobos</a>, California, near where he lived for many years.</p>\r\n"
```
Comment (birkland) [inline tests/10-migration-backend-tests/testcafe/migrations/persons-01.csv]: The description text as it appears in Drupal is kind of rough.  Does this point to a bigger issue?  Is it not possible to get decent formatted text into Drupal via migrations/CSVs, or is this a matter of "not worth it for a test"?

File: codebase/config/sync/core.extension.yml [hunk 3]
```diff
@@ -58,6 +59,7 @@ module:
   islandora_text_extraction: 0
   islandora_text_extraction_defaults: 0
   islandora_video: 0
+  jsonapi: 0
```
Comment (birkland) [inline codebase/config/sync/core.extension.yml:62]: This enables JSONAPI for all environments.  I think that is fine/useful in this case regardless of whether it's directly used by anything in production or not, as it gives flexibility for interacting with Drupal.

**Review by birkland — APPROVED** [review]
I'm going to approve this prerequisite work, as it works as advertised and is part of a series of related work.  I did notice some issues that may or may not need to be addressed down the road:

* After running the migration tests, I see error messages in the logs, like
```
 Path: /migrate_source_ui. Drupal\Core\Http\Exception\CacheableAccessDeniedHttpException: The 'access migrate source ui' permission is required. in Drupal\Core\Routing\AccessAwareRouter->checkAccess() (line 117 of /var/www/drupal/web/core/lib/Drupal/Core/Routing/AccessAwareRouter.php).
```

and 
```
Notice: Undefined index: keyname in Drupal\idc_migration\Plugin\migrate\process\Deepen->transform() (line 61 of /var/www/drupal/web/modules/contrib/idc_migration/src/Plugin/migrate/process/Deepen.php) 
```

* The migrated repository item is not "complete" with respect to "required" fields being present.  I'm aware that migration doesn't have validation; the net effect is that one cannot open an item and re-save it without entering in data.  In particular, the following are missing 
  * Subject
  * Date available
  * Access rights
  * Copyright and use
  
 * You do note that some text is kind of messed up.  
image
I added an inline comment asking if this represents a fundamental flaw, or is fixable.  There are extraneous backslashes, and (in firefox, at least), the links don't work.  It would be nice to be able to characterize the limitations of text formatting via migrations, or document what's needed in order to get "good" results.

**Comment by emetsger** [comment]
@birkland 

> * After running the migration tests, I see error messages in the logs, like
> 
> ```
>  Path: /migrate_source_ui. Drupal\Core\Http\Exception\CacheableAccessDeniedHttpException: The 'access migrate source ui' permission is required. in Drupal\Core\Routing\AccessAwareRouter->checkAccess() (line 117 of /var/www/drupal/web/core/lib/Drupal/Core/Routing/AccessAwareRouter.php).
> ```
> 
> and
> 
> ```
> Notice: Undefined index: keyname in Drupal\idc_migration\Plugin\migrate\process\Deepen->transform() (line 61 of /var/www/drupal/web/modules/contrib/idc_migration/src/Plugin/migrate/process/Deepen.php) 
> ```

I do not see these errors in my `drupal` container log after running tests for this PR, so I'm not sure what is going on in this case.

>   image
>   I added an inline comment asking if this represents a fundamental flaw, or is fixable.  There are extraneous backslashes, and (in firefox, at least), the links don't work.  It would be nice to be able to characterize the limitations of text formatting via migrations, or document what's needed in order to get "good" results.

As of this moment I haven't spent much time on investigating the formatting of descriptions.  My understanding/impression is that the description can contain basic HTML markup if it wishes to display a formatted description.  I think the Weston description simply needs to have its HTML cleaned up in the migration csv. The Adams description is an accurate representation of what a properly formatted description would look like (as represented in the CSV and represented in the UI).

**Comment by emetsger** [comment]
@birkland 16a7763 should clean up the descriptions!

(And I take back my earlier statement - the Adams description had errors, along with Weston)

**Pull request merged**